### PR TITLE
[cmake] WIP

### PR DESCRIPTION
**Pull request by dg0yt**
Adressing baseline regression: The Windows DLL triplets fail to build downstream cmake.
~~~
[527/528] C:\Windows\system32\cmd.exe /C "cd . && D:\downloads\tools\cmake-3.29.2-windows\cmake-3.29.2-windows-i386\bin\cmake.exe -E vs_link_exe --intdir=Source\QtDialog\CMakeFiles\cmake-gui.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\mt.exe --manifests D:\b\cmake\src\17503183d2-2d56d3dcb1.clean\Source\cmake.version.manifest -- C:\PROGRA~1\MICROS~1\2022\ENTERP~1\VC\Tools\MSVC\1438~1.331\bin\Hostx64\x64\link.exe  Source\QtDialog\CMakeFiles\CMakeGUIQRCLib.dir\qrc_CMakeSetup.cpp.obj Source\CMakeFiles\CMakeVersion.dir\CMakeVersion.rc.res Source\QtDialog\CMakeFiles\cmake-gui.dir\CMakeGUIExec.cxx.obj Source\QtDialog\CMakeFiles\cmake-gui.dir\CMakeSetup.rc.res  /out:bin\cmake-gui.exe /implib:bin\cmake-gui.lib /pdb:bin\cmake-gui.pdb /version:0.0 /machine:x64 -stack:10000000 /nologo    /debug /INCREMENTAL /subsystem:windows  Source\QtDialog\CMakeGUIMainLib.lib  Source\QtDialog\CMakeGUILib.lib  Source\CMakeLib.lib  Source\kwsys\cmsys.lib  Utilities\std\cmstd.lib  D:\installed\x64-windows\debug\lib\libexpatd.lib  D:\installed\x64-windows\debug\lib\zlibd.lib  D:\installed\x64-windows\debug\lib\archive.lib  D:\installed\x64-windows\debug\lib\libcurl-d.lib  D:\installed\x64-windows\debug\lib\jsoncpp.lib  D:\installed\x64-windows\debug\lib\uv.lib  psapi.lib  user32.lib  advapi32.lib  iphlpapi.lib  ws2_32.lib  dbghelp.lib  ole32.lib  Utilities\cmlibrhash\cmlibrhash.lib  rpcrt4.lib  crypt32.lib  D:\installed\x64-windows\debug\lib\Qt6Widgetsd.lib  D:\installed\x64-windows\debug\lib\Qt6Guid.lib  D:\installed\x64-windows\debug\lib\Qt6Cored.lib  mpr.lib  userenv.lib  D:\installed\x64-windows\debug\lib\Qt6EntryPointd.lib  shell32.lib  d3d11.lib  dxgi.lib  dxguid.lib  d3d12.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
[527/528] C:\Windows\system32\cmd.exe /C "cd /D D:\b\cmake\x64-windows-dbg && D:\b\cmake\x64-windows-dbg\bin\cmake.exe -P cmake_install.cmake"
FAILED: CMakeFiles/install.util 
C:\Windows\system32\cmd.exe /C "cd /D D:\b\cmake\x64-windows-dbg && D:\b\cmake\x64-windows-dbg\bin\cmake.exe -P cmake_install.cmake"
~~~
The link command is executed via the installed cmake (`D:\downloads\tools\cmake-3.29.2-windows\cmake-3.29.2-windows-i386\bin\cmake.exe`),
but the failing install command is executed via the the freshly built cmake (`D:\b\cmake\x64-windows-dbg\bin\cmake.exe`). 

(From https://github.com/microsoft/vcpkg/pull/40224#issuecomment-2266412090

CMake wants it this way for native builds:
https://github.com/Kitware/CMake/blob/4aa42149d69c8e0fa542e18196463d671190710c/Source/cmGlobalGenerator.cxx#L3121-L3127

So there seems to be an issue with DLLs.

### CODE REVIEW

**Comment by dg0yt** [comment]
Also affecting D:\b\cmake\x64-windows-rel\bin\cmake.exe, i.e release config.

**Comment by dg0yt** [comment]
The deployed `cmake.exe` is able to run and report the version - the original problem only exists with the executable in the build directory.

File: scripts/test_ports/cmake/vcpkg.json [hunk 1]
```diff
@@ -6,24 +6,29 @@
...
     {
       "name": "ncurses",
       "platform": "!(windows | uwp)"
     },
-    "nghttp2",
-    "qtbase",
+    "rhash",
+    {
+      "name": "qtbase",
+      "default-features": false
```
Comment (Neumann-A) [inline scripts/test_ports/cmake/vcpkg.json:27]: that is probably wrong.  needs at least widgets and gui

**Comment by Neumann-A** [comment]
What is the link line for cmake.exe itself ?

> the original problem only exists with the executable in the build directory.

With ADD_BIN_TO_PATH it should run just fine. I did this intentionally so that I am sure that the build cmake actually works without having to test it manually.

**Comment by dg0yt** [comment]
> What is the link line for cmake.exe itself ?

CI logs:
~~~
[512/518] C:\Windows\system32\cmd.exe /C "cd . && D:\downloads\tools\cmake-3.29.2-windows\cmake-3.29.2-windows-i386\bin\cmake.exe -E vs_link_exe --intdir=Source\CMakeFiles\cmake.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100226~1.0\x64\mt.exe --manifests D:\b\cmake\src\17503183d2-7dbde7966f.clean\Source\cmake.version.manifest -- C:\PROGRA~1\MICROS~1\2022\ENTERP~1\VC\Tools\MSVC\1438~1.331\bin\Hostx64\x64\link.exe  Source\CMakeFiles\CMakeVersion.dir\CMakeVersion.rc.res Source\CMakeFiles\cmake.dir\cmakemain.cxx.obj Source\CMakeFiles\cmake.dir\cmcmd.cxx.obj  /out:bin\cmake.exe /implib:bin\cmake.lib /pdb:bin\cmake.pdb /version:0.0 /machine:x64 -stack:10000000 /nologo /DEBUG /INCREMENTAL:NO /OPT:REF /OPT:ICF   /subsystem:console  Source\CMakeLib.lib  Source\kwsys\cmsys.lib  Utilities\std\cmstd.lib  D:\installed\x64-windows\lib\libexpat.lib  D:\installed\x64-windows\lib\zlib.lib  D:\installed\x64-windows\lib\archive.lib  D:\installed\x64-windows\lib\libcurl.lib  D:\installed\x64-windows\lib\jsoncpp.lib  D:\installed\x64-windows\lib\uv.lib  psapi.lib  user32.lib  advapi32.lib  iphlpapi.lib  userenv.lib  ws2_32.lib  dbghelp.lib  ole32.lib  shell32.lib  D:\installed\x64-windows\lib\rhash.lib  rpcrt4.lib  crypt32.lib  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cd ."
~~~

> 
> > the original problem only exists with the executable in the build directory.
> 
> With ADD_BIN_TO_PATH it should run just fine. I did this intentionally so that I am sure that the build cmake actually works without having to test it manually.

I agree. However it fails, and I don't have a local msvc installation to investigate the difference.

File: scripts/test_ports/cmake/vcpkg.json (hunk 1 above)
Comment (dg0yt) [reply scripts/test_ports/cmake/vcpkg.json:27]: True. And doesn't matter for CI. (What I really want is a feature, but it is not a priority.)

**Comment by Neumann-A** [comment]
Release only triplet:
```
E:\vcpkg_folders\various\buildtrees\cmake\x64-windows-release-rel\bin>set PATH=E:\vcpkg_folders\various\installed\x64-windows-release\bin;%PATH%

E:\vcpkg_folders\various\buildtrees\cmake\x64-windows-release-rel\bin>cmake --version
cmake version 3.22.2

CMake suite maintained and supported by Kitware (kitware.com/cmake).
```

No problem....

**Comment by dg0yt** [comment]
So it is only in CI, and only before app-local deployment?

**Comment by dg0yt** [comment]
And did you pull in all transitive dependencies which are pulled in via test ports?

For example, with curl, you also need vcpkg-ci-curl. Which got a number of additions recently.

**Comment by Neumann-A** [comment]
> And did you pull in all transitive dependencies which are pulled in via test ports?

I only pulled in the minimal required deps. x64-windows this way also worked. I can test with more but in the end it is probably a release/debug problem.

**Comment by Neumann-A** [comment]
./vcpkg.exe install --overlay-ports=scripts/test_ports/ cmake vcpkg-ci-curl --host-triplet=x64-windows

worked as expected

**Comment by Neumann-A** [comment]
Looking at the link line I would guess libarchive might be to blame if libcurl isnt the one to blame.

**Comment by Neumann-A** [comment]
Hmm building with libarchive[*] also works.

**Comment by dg0yt** [comment]
#40224 passed CI now. The actual `cmake` issue probably involves an installation order problem.

**Comment by dg0yt** [comment]
Probably symptom of https://github.com/microsoft/vcpkg/issues/40414, fixed by https://github.com/microsoft/vcpkg/pull/40473.

**Comment by dg0yt** [comment]
Most changes might still be useful.

**Pull request closed without merging**